src/crypto/mod.rs

```diff
@@ -0,0 +1,293 @@
+mod error;
+
+use pairing::{CurveAffine, CurveProjective, Engine, Field, PrimeField};
+use rand::{ChaChaRng, Rand, Rng, SeedableRng};
+use ring::digest;
+
+use self::error::{ErrorKind, Result};
+
+/// Returns a hash of the given message in `G2`.
+pub fn hash_g2<E, M>(msg: M) -> E::G2
+where
+    E: Engine,
+    <E as Engine>::G2: Rand,
+    M: AsRef<[u8]>,
+{
+    let digest = digest::digest(&digest::SHA512, msg.as_ref());
+    // The `pairing` crate's `G2` implements `Rand`. We initialize a seedable RNG with the SHA512
+    // digest, and use it to generate the element.
+    let mut msg_u32: Vec<u32> = Vec::with_capacity((digest.as_ref().len() + 3) / 4);
+    for chunk in digest.as_ref().chunks(4) {
```

Comment:
If I'm reading this correctly, you are serializing a byte-array into a `Vec<u32>` here. The `byteorder` crate has a convenient function for doing just this, if you're not aware of it already, it would allow you simplify this code block into:

```rust
use byteorder::{BigEndian, ByteOrder};
...
for chunk in digest.as_ref().chunks(4) {
    let word = BigEndian::read_u32(chunk);
    msg_u32.push(word);
}

```

Reply:
Unfortunately that panics if `word.len() < 4`, so we'd have to add padding; not sure if the code would still be as concise with that?

Reply:
Done.